onnx/common/tensor.h

```diff
@@ -139,6 +151,219 @@ struct Tensor final {
+  if (owner->is_raw_data())  {                                                 \
+    for (size_t i = 0; i < raw_data_.size(); i += sizeof(type))  {             \
+        vals.push_back(*((const type*)(owner->raw().c_str() + i)));            \
+    }                                                                          \
+    owner##_data_ptr = (type*) vals.data();                                    \
+  } else {                                                                     \
+    owner##_data_ptr = (type*) owner->vec().data();                            \
+  }                                                                            \
+
+
+#define SET_RAW_DATA(ptr)                                                      \
+  if (is_raw_data_)  {                                                         \
+    raw_data_.assign((const char*) ptr, raw_data_.size());                     \
+  }                                                                            \
+
+
+
+template<typename F, typename T>
+inline void Tensor::bin_func(F f, T* ptr, const T* a_ptr) {
+  int64_t num_elements = std::accumulate(sizes_.begin(), sizes_.end(),
+                                        (int64_t) 1, std::multiplies<int64_t>());
+  for (int64_t i = 0; i < num_elements; ++i) {
+    ptr[i] = f((T) ptr[i], (T) a_ptr[i]);
+  }
+  SET_RAW_DATA(ptr)
+}
+
+template<typename F, typename T>
+inline void Tensor::un_func(F f, T* ptr)  {
+  int64_t num_elements = std::accumulate(sizes_.begin(), sizes_.end(),
+                                        (int64_t) 1, std::multiplies<int64_t>());
+  for (int64_t i = 0; i < num_elements; ++i) {
+    ptr[i] = f((T) ptr[i]);
+  }
+  SET_RAW_DATA(ptr)
+}
+
+template<typename T>
+inline void Tensor::scale_dim(T* ptr, const T* s_ptr)  {
+  int64_t elems_per_first_dim = std::accumulate(sizes_.begin() + 1, sizes_.end(),
+                                              (int64_t) 1, std::multiplies<int64_t>());
+  int64_t first_dim_size = sizes_[0];
+  int64_t counter = 0;
+  for (int64_t i = 0; i < first_dim_size; ++i)  {
+    for (int64_t j = 0; j < elems_per_first_dim; ++j) {
+      ptr[counter++] *= s_ptr[i];
+    }
+  }
+  SET_RAW_DATA(ptr)
+}
+
+#define CALL_BIN_FUNC(type, vec, f)                                            \
+  DATA(this, type, vec)                                                        \
+  CONST_DATA(a, type, vec)                                                     \
+  bin_func<f<type>, type>(f<type>(), this_data_ptr, a_const_data_ptr);         \
+
+
+
+#define APPLY_BINARY_FUNCTION(op_name, f)                                      \
+  inline void Tensor::op_name(const Tensor& a_tensor) {                        \
+    const Tensor* a = &a_tensor;                                               \
+    if (a->elem_type() != elem_type_) {                                        \
+      throw("Type of tensors do not match");                                   \
```

Comment:
create a dedicated exception subclass for optimizer like here https://github.com/onnx/onnx/blob/master/onnx/checker.h#L13-L33 and throw that instead of general exception

Reply:
added